### PR TITLE
WattTime Model Implementation

### PR DESCRIPTION
**Pull request by gnanakeethan**
### Types of changes
- [ ] Enhancement (project structure, spelling, grammar, formatting)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### A description of the changes proposed in the Pull Request
Implementation of WattTime Model for calculating the energy intensity in gCO2e/KWh.

### CODE REVIEW

File: docs/implementations/Readme.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
 # Implementations
 
 [CCF](./ccf.md) - Cloud Carbon Footprint Impact Plugin Documentation
+
+[WattTime](./watttime.md) - WattTime Impact Plugin Documentation
```
Comment (narekhovhannisyan) [inline docs/implementations/Readme.md:5]: you can add all other implementation links here too

File: src/lib/watt-time/index.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      if (username.startsWith('ENV_')) {
+        username = process.env[username.slice(4)] ?? '';
+      }
+      if (password.startsWith('ENV_')) {
+        password = process.env[password.slice(4)] ?? '';
+      }
+      if (username === '' || password === '') {
+        throw new Error('Missing username or password & token');
+      }
+      const tokenResponse = await axios.get(this.baseUrl + '/login', {
```
Comment (narekhovhannisyan) [inline src/lib/watt-time/index.ts:32]: url can be presented with template literal which is easier to read

`${this.baseUrl}/login`

**Review by narekhovhannisyan — APPROVED** [review]

File: docs/implementations/Readme.md (hunk 1 above)
Comment (gnanakeethan) [reply docs/implementations/Readme.md:5]: I will add a separate PR with others.

**Review by jawache — CHANGES_REQUESTED** [review]
@gnanakeethan
- Very light on documentation. I've given the same feedback on lack of documentation many times before, it's not going to change. Please can you assume that for every issue ongoing that documentation is going to be a strong requirement. Both the readme and the functions, this one has one on authentication which we've never used before and has some specific functionality with respect to auth and none of it is documented as far as I can see.
- Also the original issue had requirements on how to use the WattTime API so we don't break it's rate limits. We need to think much more carefully about this approach, please re-read the original issue and my comments in this PR.

File: src/lib/watt-time/index.test.ts
[the remark this answers is not in the document]
Comment (gnanakeethan) [reply src/lib/watt-time/index.test.ts:1]: @jawache I am not testing the response from the API in the unit-test scenario. I am mocking the response. 

I will run it via the RIMPL locally to test this.

File: src/lib/watt-time/index.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    }
+  }
+
+  async calculate(
+    observations: object | object[] | undefined
+  ): Promise<object | object[]> {
+    if (!Array.isArray(observations)) {
+      throw new Error('observations should be an array');
+    }
+    observations = await Promise.all(
```
[the remark this answers is not in the document]
Comment (gnanakeethan) [reply src/lib/watt-time/index.ts:48]: @jawache that might require extensive processing before/after we obtain the data from Watt-Time API. 

I would suggest we create an intermediary mechanism which groups the data to 5 min intervals / larger intervals before we use the watt-time model. 

Also we need to consider about the arbitrary durations this model has to execute.

File: src/lib/watt-time/index.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+import * as dayjs from 'dayjs';
+
+export class WattTimeGridEmissions implements IImpactModelInterface {
+  authParams: object | undefined = undefined;
+  token = '';
+  staticParams: object | undefined;
+  name: string | undefined;
+  baseUrl = 'https://api2.watttime.org/v2';
+
+  async authenticate(authParams: object): Promise<void> {
```
[the remark this answers is not in the document]
Comment (gnanakeethan) [reply src/lib/watt-time/index.ts:13]: added in documentation.

File: src/lib/watt-time/index.test.ts
[the remark this answers is not in the document]
Comment (gnanakeethan) [reply src/lib/watt-time/index.test.ts:1]: added some custom errors to be thrown when there is no data.

**Pull request merged**